### PR TITLE
Content Transformers

### PR DESCRIPTION
**Pull request by Shemnei**
Transforms run once for each defined dotfile during the deploy process.

They can either be specified for a whole profile, in which case each dotfile is transformed by them or they can be attached to a specific dotfile.

A transform takes the contents of a dotfile, processes it and returns a new version of the content.

The contents are either a resolved template or a non-template dotfile.

# Order of execution

1) Content transformers specified in the profile
2) Content transformers specified in the specific dotfile

# Available Transformers

## LineTerminator

The `LineTerminator` transformer replaces all occurrences of one kind of line terminator with a other one.

### Options

- `LF` will replace all `\r\n` with a single `\n` (windows style -> unix style)
- `CRLF` will replace all `\n` with a `\r\n` (unix style -> windows style)

### Usage

Define the following in either the root of the profile or a specific dotfile:

```yaml
transformers:
  - LineTerminator: CRLF
```

### CODE REVIEW

**Comment by Shemnei** [comment]
bors try

File: crates/punktf-lib/src/deploy/executor.rs [hunk 1]
```diff
@@ -665,94 +676,152 @@ where
...
+			// Copy so we exec_dotfile is not referenced by this in case an error occurs.
+			let exec_transformers: Vec<_> = exec_dotfile.transformers().iter().copied().collect();
 
-			if !self.options.dry_run {
-				if let Err(err) = std::fs::write(&exec_dotfile.deploy_path(), content.as_bytes()) {
-					log::info!("{}: Failed to write content", exec_dotfile.path().display());
+			// Apply transformers.
+			// Order:
+			//   - Profile transformers
+			//   - then dotfile transformers
```
Comment (michidk) [inline crates/punktf-lib/src/deploy/executor.rs:793]: i think the difference between "dotfile" and "profile" transformers is not clear. everytring is a dotfile. so you could explain it better with per-file and per-section or sth like this. maybe we just need to introduce new nomenclaturere

File: crates/punktf-lib/src/transform.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+		fmt::Display::fmt(&self, f)
+	}
+}
+
+/// Transformer which replaces line termination characters with either unix
+/// style (`\n`) or windows style (`\r\b`).
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, serde::Serialize, serde::Deserialize)]
+pub enum LineTerminator {
+	/// Replaces all occurrences of `\r\n` with `\n` (unix style).
+	Lf,
```
Comment (michidk) [inline crates/punktf-lib/src/transform.rs:58]: shouldn't that be upper case since each char stands for its own word? e.g LF and CRLF

**Review by michidk — CHANGES_REQUESTED** [review]
Minor changes required

**Comment by michidk** [comment]
bors try

File: crates/punktf-lib/src/transform.rs (hunk 2 above)
Comment (Shemnei) [reply crates/punktf-lib/src/transform.rs:58]: I thought about it but rust for example has the `Cow` struct (Copy on write) or `Arc`. I also tought about `Lf` and `CrLf` but i think thats the worst of them all.
I think your solution is probably the best one.

Comment (Shemnei) [reply crates/punktf-lib/src/transform.rs]: Serde also allows us to set an/multiple aliases for the enum variant. So we also could allows e.g. `crlf` `CrLf` if we wanted to. I dont know it that would help or confuse people though.

**Review by michidk — APPROVED** [review]

**Comment by michidk** [comment]
bors try

**Comment by Shemnei** [comment]
Tank you very much for the review 👍.
bors r+

**Pull request merged**